models/domain.go

```diff
@@ -35,20 +36,55 @@ type scanStore interface {
 
 // IsQueueable returns true if a domain can be submitted for validation and
 // queueing to the STARTTLS Everywhere Policy List.
-func (d *Domain) IsQueueable(db scanStore, list policyList) (bool, string) {
-	// Check if successful scan occurred.
+func (d *Domain) IsQueueable(db scanStore, list policyList) (bool, string, Scan) {
 	scan, err := db.GetLatestScan(d.Name)
 	if err != nil {
 		return false, "We haven't scanned this domain yet. " +
 			"Please use the STARTTLS checker to scan your domain's " +
-			"STARTTLS configuration so we can validate your submission"
+			"STARTTLS configuration so we can validate your submission", scan
 	}
 	if scan.Data.Status != 0 {
-		return false, "Domain hasn't passed our STARTTLS security checks"
+		return false, "Domain hasn't passed our STARTTLS security checks", scan
 	}
-	// Check to see it's not already on the Policy List.
 	if list.HasDomain(d.Name) {
-		return false, "Domain is already on the policy list!"
+		return false, "Domain is already on the policy list!", scan
 	}
-	return true, ""
+	if d.MTASTSMode != "" && !scan.SupportsMTASTS() {
+		return false, "Domain does not correctly implement MTA-STS.", scan
+	} else if !subset(d.MXs, scan.Data.PreferredHostnames) {
+		return false, "Domain is not valid for the supplied hostnames.", scan
+	}
+	return true, "", scan
+}
+
+// PopulateFromScan updates a Domain's fields based on a scan of that domain.
+func (d *Domain) PopulateFromScan(scan Scan) {
+	// We should only trust MTA-STS info from a successful MTA-STS check.
+	if scan.Data.MTASTSResult != nil && scan.SupportsMTASTS() {
+		d.MTASTSMode = scan.Data.MTASTSResult.Mode
+		// If the domain's MXs are missing, we can take them from the scan's
+		// PreferredHostnames, which must be a subset of those listed in the
+		// MTA-STS policy file.
+		if len(d.MXs) == 0 {
+			d.MXs = scan.Data.PreferredHostnames
+		}
```

Comment:
We should populate this from the actual MTA-STS policy file result!

Reply:
Okay cool. My understanding was that "preferred hostnames" are all the hostnames listened in their DNS record that we were able to connect to, and that the hostnames listed in a valid MTA-STS record would be a superset of those. Is the idea that we want to capture hosts in the MTA-STS record that they might want to use in the future?

Reply:
Oh but I guess the policy is a little different from a list of actual hostnames...